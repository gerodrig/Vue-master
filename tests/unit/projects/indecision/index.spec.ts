import { vi, beforeEach, describe, expect, it, Mock } from "vitest";
import { shallowMount } from "@vue/test-utils";

import Indecision from "../../../../src/projects/indecision/index.vue";


describe('Indecision App in Vue 3', () => {
    
    let wrapper: any;
    let clgSpy: any;

    global.fetch = vi.fn( () => Promise.resolve({
        json: () => Promise.resolve({
            answer: 'Yes',
            forced: false,
            image: 'https://yesno.wtf/assets/yes/1.gif'
        }),
    }) ) as any;

    beforeEach(() => {
        wrapper = shallowMount(Indecision);

        clgSpy = vi.spyOn(console, 'log');

        //clear all mocks
        vi.clearAllMocks();
    });

    it('Must match snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('typing in input must not trigger fetch', async () => {

        const getAnswerSpy = vi.spyOn(wrapper.vm, 'getAnswer');

        const input = wrapper.find('input');
        await input.setValue('test');
        
        expect( clgSpy ).toHaveBeenCalledTimes(1);
        expect( getAnswerSpy ).not.toHaveBeenCalled();
        
    });

    it('Must fetch data from API when "?" is entered', async () => {
            
            const getAnswerSpy = vi.spyOn(wrapper.vm, 'getAnswer');
    
            const input = wrapper.find('input');
            await input.setValue('I am rich?');
            
            expect( getAnswerSpy ).toHaveBeenCalled();
            
    });

    it('tests on getAnswer', async () => {
        await wrapper.vm.getAnswer();

        const img = wrapper.find('img');

        expect(img.exists()).toBe(true);
        expect(img.attributes('src')).toBe('https://yesno.wtf/assets/yes/1.gif');
        expect(wrapper.vm.answer).toBe('Yes');
    });

    it('tests on getAnswer - handling error when API fetch fails', async () => {
        global.fetch = vi.fn( () => Promise.reject('API is down') ) as any;

        await wrapper.vm.getAnswer();

        const img = wrapper.find('img');

        expect(img.exists()).toBeFalsy();
        expect(wrapper.vm.answer).toBe('Sorry, something went wrong');
    });
});