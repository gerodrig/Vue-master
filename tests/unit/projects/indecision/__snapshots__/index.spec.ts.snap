// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Indecision App in Vue 3 > Must match snapshot 1`] = `
"<header-stub title=\\"Decision Maker\\"></header-stub>
<!--v-if-->
<div class=\\"fixed top-0 left-0 w-full h-full bg-black bg-opacity-40 -z-10\\"></div>
<div class=\\"container z-10 pt-40 mx-auto text-center\\"><input type=\\"text\\" placeholder=\\"Ask a question...\\" class=\\"w-full px-4 py-2 text-gray-800 bg-white border-none rounded-lg shadow-lg md:w-96\\">
  <p class=\\"mt-2 text-gray-300 lg:text-3xl md:text-sm\\">Remember to end your question with (?)</p>
</div>
<!--v-if-->"
`;
